hs-ucan/library/Web/UCAN/Capabilities.hs

```diff
@@ -0,0 +1,304 @@
+attempt failable f =
+  case failable of
+    Right a -> f a
+    Left e  -> return $ Left e
+
+
+resolveToken :: Resolver m => UCAN.Witness -> m (Either Error Text)
+resolveToken = \case
+  UCAN.Nested text -> return $ Right text
+  UCAN.Reference cid -> do
+    resolved <- resolve cid
+    case resolved of
+      Left err -> return $ Left $ ResolveError err
+      Right bs -> return $ Right $ decodeUtf8Lenient bs
+
+
+parseJSONString ::
+  ( FromJSON fct
+  , FromJSON res
+  , FromJSON abl
+  ) => Text -> Either Error (UCAN fct res abl)
+parseJSONString token =
+  case fromJSON $ String token of
+    Error e       -> Left $ ParseError e
+    Success proof -> Right proof
+
+
+-- checkProofLayer ::
+--   ( DelegationSemantics res
+--   , DelegationSemantics abl
+--   , Eq res
+--   , Eq abl
+--   ) => Proof fct res abl -> Bool
+-- checkProofLayer (ProofAuthorization resource ability UCAN.UCAN{ claims } step) =
+--   case step of
+--     ProofByParenthood ->
+--       -- parenthood
+--       orig == UCAN.sender ucanClaims
+--       -- Capability integrity
+--       && any (== capability) (UCAN.attenuation ucanClaims)
+--       -- Time bounds
+--       && expiration <= UCAN.expiration ucanClaims
+--       && fromMaybe True ((>=) <$> notBefore <*> UCAN.notBefore ucanClaims)
+
+--     ProofByDelegation parentProof ->
+--       -- Proof integrity
+--           UCAN.sender ucanClaims
+--       == UCAN.receiver witnessClaims
+--       -- Capability integrity
+--       && any (== capability) (UCAN.attenuation ucanClaims)
+--       -- Delegation ability
+--       && checkCapabilityDelegation capability (UCAN.sender ucanClaims) witnessClaims
+--       -- Time bounds
+--       && expiration <= UCAN.expiration ucanClaims
+--       && fromMaybe True ((>=) <$> notBefore <*> UCAN.notBefore ucanClaims)
+
+-- checkAuthProof :: ProofAuthentication fct res abl -> Bool
+-- checkAuthProof (ProofAs did scope UCAN.UCAN{ claims } parentProof) =
+--   -- Proof integrity
+--   UCAN.sender claims
+
+
+-- checkCapabilityDelegation ::
+--   ( DelegationSemantics res
+--   , DelegationSemantics abl
+--   ) => UCAN.Capability res abl -> DID -> UCAN.Claims fct res abl -> Bool
+-- checkCapabilityDelegation capability sender witnessClaims =
+--   case capability of
+--     UCAN.CapProofRedelegation _ ->
+--       True
+
+--     UCAN.CapOwnedResources ownedRes@(UCAN.OwnedResources did _) ->
+--       did == sender || any (`canDelegate` ownedRes) (filterOwnedResources =<< UCAN.attenuation witnessClaims)
+
+--     UCAN.CapResource resource ability ->
+--       any (`canDelegate` (resource, ability)) (filterResources =<< UCAN.attenuation witnessClaims)
+--   where
+--     filterOwnedResources = \case
+--       UCAN.CapOwnedResources ownedResources -> [ownedResources]
+--       _                                     -> []
+
+--     filterResources = \case
+--       UCAN.CapResource resource ability -> [(resource, ability)]
+--       _                                 -> []
+
```

Comment:
Yeah, this is all WIP, sorry. The idea is to write a `checkProof` function which can take `Proof`s, and verify their integrity, mostly because this is useful for testing (but possibly for something else?). Maybe I'll remove that / move that to tests.